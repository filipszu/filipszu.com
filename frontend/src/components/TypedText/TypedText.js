import classes from './TypedText.module.css';
import React, { useEffect, useState } from 'react';

const TypedText = (props) => {
    const requestRef = React.useRef();
    const previousTimeRef = React.useRef();
    const lastAnimationTickRef = React.useRef();
    const isDoneAnimating = React.useRef();
    const [isAnimating, setIsAnimating] = useState(false);
    const [typedText, setTypedText] = useState('');

    const type = () => {
        setTypedText(prevTypedText => { 
            const newLength = prevTypedText.length+1;
            const newValue = props.children.substr(0, newLength);
            if(newLength === props.children.length){
                isDoneAnimating.current = true;
            }
            return newValue;
        });
    };

    const animate = time => {
        if (previousTimeRef.current !== undefined) {
            
            setIsAnimating(prevAnimating => {
                if(!prevAnimating && time >= props.delay){
                    lastAnimationTickRef.current = time;
                    type();
                    return true;
                }
    
                if(prevAnimating && time - lastAnimationTickRef.current >= props.interval){
                    lastAnimationTickRef.current = time;
                    type();
                }

                return prevAnimating;
            });
          }

          previousTimeRef.current = time;
          if(!isDoneAnimating.current){
            requestRef.current = requestAnimationFrame(animate);
          }
    };

    useEffect(() => {
        requestRef.current = requestAnimationFrame(animate);
        return () => cancelAnimationFrame(requestRef.current);
    }, []);
    
    const cursorClasses = [classes.Pulse, classes.TextCursor].join(' ');
    return (
        <div className={props.className}>
            <p>
                <span dangerouslySetInnerHTML={{__html: typedText}}></span><span className={cursorClasses}>â–®</span>
            </p>
        </div>
    );
};

export default TypedText;