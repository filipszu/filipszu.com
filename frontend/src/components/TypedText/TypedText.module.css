.TextCursor{
	padding-left:3px;
	color: #FFFFFF;
	padding-bottom: 6px;
}

.Pulse{
	animation: blink .5s infinite;
}

@keyframes blink {
	0% {
	  opacity: 0;
	}
	50% {
	  opacity: 1;
	}
	100%{
		opacity: 0;
	}
}