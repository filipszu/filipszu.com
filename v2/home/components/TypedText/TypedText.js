import classes from './TypedText.module.css';
import { useState } from 'react';
import useAnimationFrame from '../../hooks/useAnimationFrame';

const TypedText = (props) => {
    const [typedText, setTypedText] = useState('');

    const type = time => {
        let interval = parseInt(props.interval);
        if(time >= props.delay) {
            setTypedText(prevTypedText => { 
                const newLength = prevTypedText.length+1;
                const newValue = props.children.substr(0, newLength);
                if(newLength === props.children.length){
                    interval = 0;
                }
                return newValue;
            });
        }
        return interval;
    };
    const cursorClasses = [classes.Pulse, classes.TextCursor].join(' ');

    useAnimationFrame(type);

    return (
        <div className={props.className}>
            <p>
                <span dangerouslySetInnerHTML={{__html: typedText}}></span><span className={cursorClasses}>â–®</span>
            </p>
        </div>
    );
};

export default TypedText;