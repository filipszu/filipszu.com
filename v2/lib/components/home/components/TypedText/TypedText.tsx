import React from "react";
import classes from "./TypedText.module.css";
import { useState } from "react";
import useAnimationFrame, {
  useAnimationFrameCallback,
} from "../../hooks/useAnimationFrame";

export interface TypedTextProps {
  text: string;
  interval: number;
  delay: number;
  className?: string;
}

const TypedText = (props: TypedTextProps) => {
  const [typedText, setTypedText] = useState("");

  const type: useAnimationFrameCallback = (time) => {
    let interval = props.interval;
    if (time >= props.delay && props.text !== (null || undefined)) {
      setTypedText((prevTypedText) => {
        const newLength = prevTypedText.length + 1;
        const newValue = props.text.substr(0, newLength);
        if (newLength === props.text.length) {
          interval = 0;
        }
        return newValue;
      });
    }
    return interval;
  };
  const cursorClasses = [classes.Pulse, classes.TextCursor].join(" ");

  useAnimationFrame(type);

  return (
    <div className={props.className}>
      <p>
        <span dangerouslySetInnerHTML={{ __html: typedText }}></span>
        <span className={cursorClasses}>â–®</span>
      </p>
    </div>
  );
};

export default TypedText;
