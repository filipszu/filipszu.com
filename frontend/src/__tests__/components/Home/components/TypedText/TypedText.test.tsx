import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import TypedText, {
  TestID,
} from "../../../../../components/Home/components/TypedText/TypedText";

const hardcodedCursor = "â–®";
const mockedText = "The mocked text.";

describe(`TypedText component responsibble for animating the message`, () => {
  test(`It will render correctly and match the snapshot.`, () => {
    const tree = renderer
      .create(<TypedText text={mockedText} delay={500} interval={10} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test(`[Given] it will render [Then] it will use the appropriate test-id: ${TestID}`, () => {
    render(<TypedText text={mockedText} delay={500} interval={10} />);
    const el = screen.getByTestId(TestID);
    expect(el).toBeInTheDocument();
  });

  test(`[Given] it will render [Then] it will render the hardcoded cursor: ${hardcodedCursor}`, () => {
    render(<TypedText text={mockedText} delay={500} interval={10} />);
    const el = screen.getByText(hardcodedCursor);
    expect(el).toBeInTheDocument();
  });

  test(`[Given] a text prop [Then] it will render it after the animation.`, async () => {
    render(<TypedText text={mockedText} delay={500} interval={10} />);
    const el = await screen.findByText(mockedText);
    expect(el).toBeInTheDocument();
  });
});
