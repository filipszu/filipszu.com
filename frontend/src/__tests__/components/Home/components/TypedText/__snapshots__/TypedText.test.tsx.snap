// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypedText component responsibble for animating the message It will render correctly and match the snapshot. 1`] = `
<div
  data-testid="szu-typedtext"
>
  <p>
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
    <span
      className="Pulse TextCursor"
    >
      â–®
    </span>
  </p>
</div>
`;
